src/test/scala/org/ergoplatform/nodeView/mempool/ErgoMemPoolSpec.scala

```diff
@@ -0,0 +1,88 @@
+package org.ergoplatform.nodeView.mempool
+
+import org.ergoplatform.ErgoBoxCandidate
+import org.ergoplatform.modifiers.mempool.ErgoTransaction
+import org.ergoplatform.nodeView.mempool.ErgoMemPool.ProcessingOutcome
+import org.ergoplatform.nodeView.state.ErgoState
+import org.ergoplatform.nodeView.state.wrapped.WrappedUtxoState
+import org.ergoplatform.utils.ErgoTestHelpers
+import org.ergoplatform.utils.generators.ErgoGenerators
+import org.scalatest.FlatSpec
+import org.scalatest.prop.PropertyChecks
+
+import scala.util.Random
+
+class ErgoMemPoolSpec extends FlatSpec
+  with ErgoGenerators
+  with ErgoTestHelpers
+  with PropertyChecks {
+
+  it should "accept valid transaction" in {
+    val (us, bh) = createUtxoState()
+    val genesis = validFullBlockWithBlockHolder(None, us, bh, Random)._1
+    val wus = WrappedUtxoState(us, bh, stateConstants).applyModifier(genesis).get
+    val txs = validTransactionsFromUtxoState(wus, Random)
+    val pool = ErgoMemPool.empty(settings)
+    txs.foreach { tx =>
+      pool.putIfValid(tx, us)._2 shouldBe ProcessingOutcome.Accepted
+    }
+  }
+
+  it should "decline already contained transaction" in {
+    val (us, bh) = createUtxoState()
+    val genesis = validFullBlockWithBlockHolder(None, us, bh, Random)._1
+    val wus = WrappedUtxoState(us, bh, stateConstants).applyModifier(genesis).get
+    val txs = validTransactionsFromUtxoState(wus, Random)
+    var pool = ErgoMemPool.empty(settings)
+    txs.foreach { tx =>
+      pool = pool.putWithoutCheck(Seq(tx))
+    }
+    txs.foreach { tx =>
+      pool.putIfValid(tx, us)._2 shouldBe ProcessingOutcome.Declined
+    }
+  }
+
+  it should "decline transactions invalidated earlier" in {
```

Comment:
I don't see the difference between current test and a previous one ("decline already contained transaction"). Where is transaction invalidation here?